lib/document.js

```diff
@@ -4582,6 +4582,27 @@ Document.prototype.getChanges = function() {
   return changes;
 };
 
+/**
+ * Returns a copy of this document with a deep clone of `_doc` and `$__`.
+ *
+ * @return {Document} a copy of this document
+ * @api private
+ * @method $clone
+ * @memberOf Document
+ * @instance
+ */
+
+Document.prototype.$clone = function() {
+  const clonedDoc = Object.assign(Object.create(Object.getPrototypeOf(this)), this);
```

Comment:
I really don't like using `Object.create()` like this, because `Object.create()` skips running the `Document` constructor. I'd try just using `const Model = this.constructor; const clonedDoc = new Model();`